### PR TITLE
Hookup non-Roslyn view to the document tracking service on a foregrou…

### PR DESCRIPTION
**Pull request by mavasani**
…nd scheduled task.

RPS traces show that we are forcing the text view creation too eagerly on the UI thread during file open. This should very likely fix VSO #[223317](https://devdiv.visualstudio.com/DevDiv/_workitems?_a=edit&id=223317&triage=true).

### CODE REVIEW

**Comment by mavasani** [comment]
@dotnet-bot retest this please

**Comment by mavasani** [comment]
Tagging @dotnet/roslyn-ide @dotnet/roslyn-perf @heejaechang

File: src/VisualStudio/Core/Def/RoslynDocumentProvider.cs [hunk 1]
```diff
@@ -40,11 +40,15 @@ protected override void OnBeforeNonRoslynDocumentWindowShow(IVsWindowFrame frame
                 return;
             }
 
-            var view = GetTextViewFromFrame(frame);
-            if (view != null)
+            // Schedule hookup of the non-Roslyn document view on foreground task scheduler.
+            InvokeBelowInputPriority(() =>
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/RoslynDocumentProvider.cs]: Ah, I see that will end up executing right away as we are on foreground thread already. I'll change this to explicitly invoke on a scheduled task.

**Comment by heejaechang** [comment]
👍

**Comment by CyrusNajmabadi** [comment]
Wait... how do this help things?  Aren't we still going to just make the UI immediately?  Don't we need the text view if the file is being opened?  @jasonmalinowski

**Comment by mavasani** [comment]
Prior to this change, we executed the code to force complete the WPF text view for non-Roslyn documents on the UI thread while the input is still pending (this code is executed when RTD is initializing the document). After this change, I verified under the debugger that we return immediately and the task is executed separately (though pretty soon afterwards).

**Comment by CyrusNajmabadi** [comment]
My point is: what benefit does this have?  The UI still needs to be opened.  So are we just kicking work down the road just to make a metric happy? :)

**Comment by mavasani** [comment]
@CyrusNajmabadi The issue is we are forcing the WpfTextView creation during RDT operation OnBeforeDocumentShow, which is against the recommended pattern. I found a better approach for this - we can completely avoid forcing WpfTextView creation during RDT operations, and hookup the listener during [DocumentProvider.TryProcessOpenForDocCookie](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/DocumentProvider.cs,324), at which point the [TextBufferAssociatedViewService](http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/TextBufferAssociatedViewService.cs,22) should already have the view for the buffer. I'll give this a try and if it works fine, will revert this change in favor of that one.

**Comment by CyrusNajmabadi** [comment]
Yes, please do that instead.  I'm not a fan of the current approach :)  Thanks!

**Comment by mavasani** [comment]
So, I tried couple of approaches for this and neither seemed to work:

1. Avoid realizing the wpf text view at all and try to hookup to the buffer. However, `EditorAdaptersFactoryService.GetDocumentBuffer` returns null for the IVsTextBuffer [here](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/DocumentProvider.cs,344).
2. Move the text view realization from [DocumentProvider.OnBeforeDocumentWindowShow](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/DocumentProvider.RunningDocTableEventsSink.cs,45) to [DocumentProvider.OnAfterAttributeChangeEx](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/DocumentProvider.cs,514) when `__VSRDTATTRIB3.RDTA_DocumentInitialized` is set, but it turns out that the latter is called before the former and hence doesn't help.

At this point, I am going to open a separate issue for the IDE team to refactor this properly and stick with the original fix as it moves the initialization out of the call stack where user input is pending.

**Comment by mavasani** [comment]
Filed https://github.com/dotnet/roslyn/issues/15153

**Comment by jasonmalinowski** [comment]
> However, EditorAdaptersFactoryService.GetDocumentBuffer returns null for the IVsTextBuffer here.

Did the debugger give a reason why? The only thing I could think of is if the IVsTextBuffer was deferred, which would mean the OnAfterAttributeChangeEx approach is exactly what we want.

> Move the text view realization from DocumentProvider.OnBeforeDocumentWindowShow to DocumentProvider.OnAfterAttributeChangeEx when __VSRDTATTRIB3.RDTA_DocumentInitialized is set, but it turns out that the latter is called before the former and hence doesn't help.

So _DocumentInitialized should only be getting fired once we're already doing everything. I agree with @CyrusNajmabadi -- the PR here still seems like we're just moving time around and we still don't understand what is going on here. Rather than deferring the work in a bug I'd prefer we attack it now and fix this right. Or we confirm that the metric being measured was too narrow (and didn't represent the actual customer experience) and we fix that as well.

**Comment by mavasani** [comment]
@jasonmalinowski helped me out here and approach 1. above works now - I had to hook up a listener if the buffer was not initialized at the point. I am going to revert this change in master and send a PR against dev15-rc2.

**Pull request merged**